### ISSUE TITLE
Pose keypoint coordinates incorrect in val() after upgrading Ultralytics — works fine after downgrading to previous version.

### ISSUE THREAD

**Issue by 201411156** [ultralytics/ultralytics#21668]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

model : yolov11n-pose
When using ultralytics 8.3.178.
And imgsz is same.

val : 2 0.154034 0.293033 0.125796 0.221968 0.154492 0.293188 0.999998
predict : 2 0.496961 0.45005 0.285424 0.570383 0.496788 0.454872 0.999999
Value is different, But when val inferenced image (val_batch0_pred.jpg ) is drawing correct bbox.

ultralytics 8.3.94
val : 2 0.462641 0.490002 0.37749 0.665694 0.463134 0.483626 0.999999
predict : 2 0.462096 0.490284 0.377429 0.665852 0.463453 0.490471 0.999998

What is the latest version's validation bbox value?
It seems like scaling correct labels.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @201411156, thank you for reporting this and for using Ultralytics 🚀 This is an automated response to help you get quicker support. An Ultralytics engineer will review and assist you soon.

Please start by checking the Docs where many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples cover common usage patterns and answers:
- Docs: https://docs.ultralytics.com/

If this is a 🐛 Bug Report (it appears to be), please provide a complete [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so we can reproduce the keypoint/bbox value differences you’re seeing between val() and predict across versions. This should include:
- Exact ultralytics version(s) tested and installation method
- Model and command(s) used for val and predict, including imgsz, conf, device, augment, batch, and any other flags
- A small sample image and expected vs actual outputs
- Any custom data.yaml or transforms being used
- The raw outputs from both val and predict

If this is a custom training ❓ Question, please share dataset samples, training/validation logs, and confirm you’ve followed our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best:
- Discord for real-time chat: https://discord.com/invite/ultralytics 🎧
- Discourse for in-depth discussions: https://community.ultralytics.com/
- Subreddit to share with the community: https://www.reddit.com/r/Ultralytics/

## Upgrade

First, please upgrade to the latest ultralytics package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a Python>=3.8 environment with PyTorch>=1.8 to confirm whether the issue persists in the latest release:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks again for the detailed version info and examples you’ve shared so far 🙏 To move forward, please add an MRE with commands, a sample image, and the exact logs/outputs for both val and predict on ultralytics 8.3.178 vs 8.3.94, so we can investigate any scaling or formatting differences in bbox/keypoint outputs.

**Comment by Y-T-G** [comment]
Can you provide the output after running this command in terminal: `yolo checks`?

Can you provide the code you ran?

**Comment by 201411156** [comment]
OK. 
But I had already downgraded my Ultralytics version when the issue first occurred. So I upgraded it back to the version where the error happens, ran the YOLO checks, and also tested val() again — the issue was reproduced exactly as before.

**### Original version**
(yolo) /workspace$ yolo checks
Ultralytics 8.3.179 🚀 Python-3.10.0 torch-2.6.0+cu124 CUDA:0 (NVIDIA A100-SXM4-80GB, 81153MiB)
Setup complete ✅ (256 CPUs, 2015.7 GB RAM, 203.1/1757.8 GB disk)

OS                     Linux-5.15.0-1083-nvidia-x86_64-with-glibc2.27
Environment            Linux
Python                 3.10.0
Install                pip
Path                   /opt/conda/envs/yolo/lib/python3.10/site-packages/ultralytics
RAM                    2015.66 GB
Disk                   203.1/1757.8 GB
CPU                    AMD EPYC 7742 64-Core Processor
CPU count              256
GPU                    NVIDIA A100-SXM4-80GB, 81153MiB
GPU count              8
CUDA                   12.4

numpy                  ✅ 2.1.1>=1.23.0
matplotlib             ✅ 3.10.5>=3.3.0
opencv-python          ❌ (not installed)>=4.6.0
pillow                 ✅ 11.3.0>=7.1.2
pyyaml                 ✅ 6.0.2>=5.3.1
requests               ✅ 2.32.4>=2.23.0
scipy                  ✅ 1.15.3>=1.4.1
torch                  ✅ 2.6.0>=1.8.0
torch                  ✅ 2.6.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.21.0>=0.9.0
tqdm                   ✅ 4.67.1>=4.64.0
psutil                 ✅ 7.0.0
py-cpuinfo             ✅ 9.0.0
pandas                 ✅ 2.3.1>=1.1.4
ultralytics-thop       ✅ 2.0.15>=2.0.0

**### Downgrade version**
(yolo) /workspace$ yolo checks
Ultralytics 8.3.94 🚀 Python-3.10.0 torch-2.6.0+cu124 CUDA:0 (NVIDIA A100-SXM4-80GB, 81153MiB)
Setup complete ✅ (256 CPUs, 2015.7 GB RAM, 203.1/1757.8 GB disk)

OS                  Linux-5.15.0-1083-nvidia-x86_64-with-glibc2.27
Environment         Linux
Python              3.10.0
Install             pip
Path                /opt/conda/envs/yolo/lib/python3.10/site-packages/ultralytics
RAM                 2015.66 GB
Disk                203.1/1757.8 GB
CPU                 AMD EPYC 7742 64-Core Processor
CPU count           256
GPU                 NVIDIA A100-SXM4-80GB, 81153MiB
GPU count           8
CUDA                12.4

numpy               ✅ 2.1.1<=2.1.1,>=1.23.0
matplotlib          ✅ 3.10.5>=3.3.0
opencv-python       ❌ (not installed)>=4.6.0
pillow              ✅ 11.3.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.4>=2.23.0
scipy               ✅ 1.15.3>=1.4.1
torch               ✅ 2.6.0>=1.8.0
torch               ✅ 2.6.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.21.0>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 7.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.3.1>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.15>=2.0.0

**Comment by 201411156** [comment]
@Y-T-G 
And I ran val both in CLI and Python code.
### **CLI :**
yolo val model=~/best.pt data=~/data.yaml imgsz=640 save_txt=True
### **Python :**
model.val(
    data=data_yaml,
    device=[0, 1, 2, 3],
    batch=64,
    imgsz=640,
    conf=0.25,
    iou=0 if is_lens_model else 0.7,
    save=True,
    save_txt=True,
    save_conf=True,
    save_json=False,
    plots=True,
    single_cls=False,
    project=f"~/",
    name=folder,
)

**Comment by huangjianaa** [comment]
the same mistake.

**Comment by Y-T-G** [comment]
Is version 8.3.167 working correctly?

**Comment by 201411156** [comment]
I’ll be test next week

**Comment by glenn-jocher** [comment]
Thanks—while you test 8.3.167, note there’s likely a regression in val() where pose save_txt writes unscaled keypoints/boxes, so normalized values differ from predict; predict explicitly scales keypoints to the original image before saving, as shown in the pose predictor construct_result docs, and val JSON does the same scaling in pred_to_json, but save_one_txt currently passes keypoints through without scaling. As a workaround, run val with save_json=True and use predictions.json, or read results.keypoints.xyn from predict; we’re preparing a fix to align val save_txt with predict.

**Comment by huangjianaa** [comment]
> 谢谢 - 当您测试 8.3.167 时，请注意 val（） 中可能存在回归，其中姿势save_txt写入未缩放的关键点/框，因此规范化值与 predict 不同;predict 在保存之前显式缩放关键点到原始图像，如 Pose Predictor construct_result 文档所示，而 val JSON 在 pred_to_json 中执行相同的缩放，但save_one_txt目前传递关键点而不缩放。作为解决方法，运行 val with save_json=True 并使用 predictions.json，或从 predict 读取 results.keypoints.xyn;我们正在准备一个修复程序，以使 val save_txt 与 predict 保持一致。

Looking forward to resolving this issue in the current version.

**Comment by glenn-jocher** [comment]
Thanks for the patience—this fix is queued and will land in the next patch; you can verify early by installing from main (pip install -U git+https://github.com/ultralytics/ultralytics), which updates PoseValidator.save_one_txt to mirror the scaling used in the PoseValidator pred_to_json reference.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this has been merged in PR #21719 by Y-T-G with contributions from Laughing-q. Thank you both!

Key changes that may resolve the issue:
- Introduces a shared scale_preds method in validation to consistently scale predictions back to original image dimensions before saving.
- Detection: Scales boxes once and reuses them for JSON/TXT, preventing mismatches or double-scaling.
- OBB: Correctly scales rotated boxes and removes redundant scaling in JSON export.
- Pose: Centralizes keypoint scaling into scale_preds for consistent TXT/JSON outputs.
- Segmentation: Moves mask resizing into scale_preds so saved masks align with original image sizes.
- Overall: Reduces duplication and improves reliability of all saved outputs, including those used by downstream tools and Ultralytics HUB.

How to test the fix:
- Install the latest main branch:
  pip install -U git+https://github.com/ultralytics/ultralytics.git@main
- Alternatively, wait for the next official PyPI release and upgrade once available:
  pip install -U ultralytics
- If a specific version is mentioned in release notes later, update to that or newer:
  pip install -U ultralytics>=VERSION

Feedback request:
- Please try the latest build and let us know if the saved TXT/JSON outputs now match original image sizes as expected. Share any remaining inconsistencies or logs if the issue persists.

Thanks for reporting this and helping improve the reliability of exports. 🙏
